aurora-is-near/rainbow-bridge
# Eth2Near-relay: improvements, add unit tests

# Unit tests for Eth2Near Relay. 
## Unit tests for contract-wrapper module
In unit tests for contract wrapper, I use only data from files and don't do any requests to the beacon chain. 

**data:**
* `execution_block_headers_kiln_1099394-1099937.json` -- JSON file with Ethereum execution block headers from slot 1099394 to slot 1099937 from Kiln Testnet. May contain some duplication. 
* `light_client_updates_kiln_10999394-1099937.json` -- JSON with light client updates from slot 10999394 to slot 1099937. 
* `next_sync_committee_133.json/134.json` -- JSON files with next sync committee for periods 133 and 134 in Kiln Testnet.

For tests  **sandbox_contract_wrapper** which implements `contract_wrapper_trait` and smart contract functions calls via sandbox were created. It can be used inside `eth_client_contract` instead of `near_contract_wrapper`. 

More tests were implemented for:
* `dao_contract`;
* `dao_eth_client_contract`;
* `eth_client_contract`.

For `dao_contract`: 
* Interaction with `MockDAOContract` and some checks inside `MockDAOContract`;
* Interaction with real DAO Contract and check that updates were submitted.

For `dao_eth_client_contract`:
* Attempts to interact with the real DAO contract. However, it makes sense to design better tests for it.

For `eth_client_contract`:
* Simple tests of all functions in `eth_client_contract`, interaction with sandbox. 

## Unit tests for eth2near-relay module
`Eth2NearRelay` tests were split into two parts: `Eth2NearRelay` and `LastSlotSerach`. 

Tests for `LastSlotSearch`: 
* Basic tests for each function;
* Check the slot searching functionality nearby slots without block;
* Check the search if the server does not respond.

Tests for `Eth2NearRelay`:
* Basic tests for each function;
* Tests with incorrect light client update;
* Tests with no response from the server;
* Tests for the light client update in the new period.

P.S. tests that fetched the beacon block state (for light client updates) are ignored by default. It's recommended not to run these tests at the same time as other tests, and it's better not to run them too often (these tests are heavy and could affect the used web3 endpoint's quota). 

# Other changes
* Errors if RPC request return incorrect JSON (typically on network problems) 
* Support logging into file
* Shot run mode: run relay till the light client update from file
* Refactoring: separate last_slot_searcher
* Refactoring: split functions into small one
* Specific error for slots without block
* Cleanup: skip_fail and return_on_fail macros
